lib/SILOptimizer/Transforms/CopyForwarding.cpp

```diff
@@ -1015,8 +1024,38 @@ bool CopyForwarding::forwardPropagateCopy() {
     // If this use cannot be analyzed, then abort.
     if (!AnalyzeUse.Oper)
       return false;
-    // Otherwise record the operand.
-    ValueUses.push_back(AnalyzeUse.Oper);
+
+    // We want to make sure that we can optimize the following:
+    //    x = alloc_stack
+    //    apply fn(x)
+    //    copy_addr y to x
+    //    apply fn2(x)
+    // into:
+    //    x = alloc_stack
+    //    apply fn(x)
+    //    apply fn2(y)
+    // But, we need to make sure that we bail in the following case:
+    //    x = alloc_stack
+    //    e = struct_element_addr x
+    //    copy_addr to x
+    //    use e
+    // To do this we check if all uses of the dest user (UserInst) dominate the
+    // current copy if the current copy is dominated by UserInst.
+    DominanceInfo domInfo(CurrentCopy->getFunction());
+    if (domInfo.dominates(UserInst, CurrentCopy)) {
+      auto *userVal = dyn_cast<SingleValueInstruction>(UserInst);
+      if (!userVal)
+        return false;
+
+      for (auto *userUse : userVal->getUses()) {
```

Comment:
This could be quadratic in the worst case. This is to address the following issue: 
```
x = alloc_stack
e = struct_element_addr x
copy_addr to x
use e
```
also, see `@testLoadDestroy` in `copyforwarding_ossa.sil`. 

I originally wanted to fix this by simply bailing on functions that had uses of dest before the copy_addr but, that would mean we couldn't optimize any of the following patterns:
```
call init fn
copy_addr
use
```

Let me know what you think the best course of action is and of course feel free to suggest alternative fixes.